### PR TITLE
Refactor thick- and thin-target code

### PR DESCRIPTION
**Pull request by samaloney**
See #40

Large scale refactoring tried to break up big parts into separate commits.

### CODE REVIEW

**Comment by samaloney** [comment]
Current comparison plots for both thin- and thick-target modes.

@binchensun thin-target looks in good shape 

[idl_sunxspec_thicktarget_comarison.pdf](https://github.com/sunpy/sunxspex/files/4370881/idl_sunxspec_thicktarget_comarison.pdf)
[idl_sunxspec_thintarget_comarison.pdf](https://github.com/sunpy/sunxspex/files/4370882/idl_sunxspec_thintarget_comarison.pdf)

**Comment by samaloney** [comment]
Still more to be done but I think this is enough for this PR 😆

File: sunxspex/emission.py [hunk 1]
```diff
@@ -0,0 +1,814 @@
...
+            a_lg = np.log10(photon_energies[P1])
+            b_lg = np.log10(np.full_like(a_lg, eelow))
+            i = np.copy(P1)
+            intsum1, ier1 = integrate_part(model=model, maxfcn=maxfcn, rerr=rerr,
+                                           photon_energies=photon_energies,
+                                           eelow=eelow, eebrk=eebrk, eehigh=eehigh, p=p, q=q, z=z,
+                                           a_lg=a_lg, b_lg=b_lg, ll=i, efd=efd)
+
+            # ier = 1 indicates no convergence.
+            if sum(ier1) > 0:
```
Comment (dpshelio) [inline sunxspex/emission.py]:
```suggestion
            if sum(ier1):
```
`> 0` is redundant here.

File: sunxspex/emission.py [hunk 2]
```diff
@@ -0,0 +1,814 @@
...
+
+    P1 = np.where(photon_energies < eelow)[0]
+    P2 = np.where(photon_energies < eebrk)[0]
+    P3 = np.where(photon_energies <= eehigh)[0]
+
+    # Part 1, below en_val[0] (usually eelow)
+    if model == 'thick-target':
+        if P1.size > 0:
+            print('Part1')
+            a_lg = np.log10(photon_energies[P1])
```
Comment (dpshelio) [inline sunxspex/emission.py]: Are you only interested on the first element/dimension of P1? (as taken from line 574)

File: sunxspex/emission.py [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
+    intsum3 = np.zeros_like(photon_energies, dtype=np.float64)
+    ier3 = np.zeros_like(photon_energies, dtype=np.float64)
+
+    P1 = np.where(photon_energies < eelow)[0]
+    P2 = np.where(photon_energies < eebrk)[0]
+    P3 = np.where(photon_energies <= eehigh)[0]
+
+    # Part 1, below en_val[0] (usually eelow)
+    if model == 'thick-target':
+        if P1.size > 0:
```
Comment (dpshelio) [inline sunxspex/emission.py]: I believe you don't need the `size` here or the `> 0`.

File: sunxspex/emission.py [hunk 4]
```diff
@@ -0,0 +1,814 @@
...
+    # TODO check units here
+    # Combine 3 parts and convert units and return
+    if model == 'thick-target':
+        DmlinO = (intsum1 + intsum2 + intsum3) * (mc2 / clight)
+        ier = ier1 + ier2 + ier3
+        return DmlinO, ier
+    elif model == 'thin-target':
+        Dmlin = (intsum2 + intsum3)
+        ier = ier2 + ier3
+        return Dmlin, ier
```
Comment (dpshelio) [inline sunxspex/emission.py]: return outside the `if` and keep the two variables `Dmlin` and `DmlinO` the same.

File: sunxspex/emission.py [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+
+    # Numerical coefficient for photo flux
+    fcoeff = (clight / (4 * np.pi * au ** 2)) / mc2 ** 2.
+
+    # Create arrays for the photon flux and error flags.
+    flux = np.zeros_like(photon_energies, dtype=np.float64)
+    iergq = np.zeros_like(photon_energies, dtype=np.float64)
+
+    if eelow >= eehigh:
+        raise ValueError('eehigh must be larger than eelow!')
```
Comment (dpshelio) [inline sunxspex/emission.py]: I'll put this check earlier on.

File: sunxspex/emission.py [hunk 6]
```diff
@@ -0,0 +1,814 @@
...
+                                                photon_energies=photon_energies[l], maxfcn=maxfcn,
+                                                rerr=rerr, eelow=eelow, eebrk=eebrk, eehigh=eehigh,
+                                                p=p, q=q, z=z, efd=efd)
+
+        flux *= fcoeff
+
+        return flux
+    else:
+        raise Warning('The photon energies are higher than the highest electron energy or not '
+                      'greater than zero')
```
Comment (dpshelio) [inline sunxspex/emission.py]: Mabye the if only for the raise and the rest outside the if to remove indentation?
```python
if not l:
    raise Warning...
flux[l], ....
```

File: sunxspex/emission.py [hunk 7]
```diff
@@ -0,0 +1,814 @@
...
+                                                maxfcn=maxfcn, rerr=rerr, eelow=eelow,
+                                                eebrk=eebrk, eehigh=eehigh, p=p, q=q, z=z,
+                                                efd=False)
+
+        flux = (fcoeff / decoeff) * flux
+
+        return flux
+    else:
+        raise Warning('The photon energies are higher than the highest electron energy or not '
+                      'greater than zero')
```
Comment (dpshelio) [inline sunxspex/emission.py]: Similar comments than the functions above.

File: sunxspex/constants.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+class Constants(metaclass=Singleton):
+    """
+    Centralised constant representation
+    """
+    def __init__(self, ref='idl'):
+        """
+
+        Parameters
+        ----------
+        ref
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/constants.py:52]: Yea I can't see there being any others I see about updating this.

File: sunxspex/emission.py [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+
+        Returns
+        -------
+        `numpy.array`
+            The electron spectrum as a function of electron energy
+        """
+        res = np.zeros_like(electron_energy)
+
+        index = np.where(electron_energy < self.eelow)
+        if index[0].size > 0:
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:112]: `np.where` always returns a non-empty tuple so I don't think this works
```
a = np.array([1,2,3])
i = np.where(a<1)
if i:
    print('here')

here
```

File: sunxspex/emission.py [hunk 10]
```diff
@@ -0,0 +1,814 @@
...
+            res[index] = 0.
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:113]: Yep this is true I could just remove the if clause altogether

File: sunxspex/emission.py [hunk 11]
```diff
@@ -0,0 +1,814 @@
...
+
+        index = np.where((electron_energy < self.eebrk) & (electron_energy >= self.eelow))
+        if index[0].size > 0:
+            res[index] = self._norm_factor * self._n0 * (self.p - 1.) \
+                         * electron_energy[index] ** (-self.p) * self.eelow ** (self.p - 1.)
+
+        index = np.where((electron_energy <= self.eehigh) & (electron_energy >= self.eebrk))
+        if index[0].size > 0:
+            res[index] = self._norm_factor * (self.q - 1.) \
+                         * electron_energy[index] ** (-self.q) * self.eebrk ** (self.q - 1.)
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:123]: Any suggestion for the names, `lt_eelow`,  `lt_eebrke`, l`e_eehigh`?

File: sunxspex/emission.py [hunk 12]
```diff
@@ -0,0 +1,814 @@
...
+    alpha = const.get_constant('alpha')
+    twoar02 = const.get_constant('twoar02')
+
+    # Numerical coefficients
+    c11 = 4.0 / 3.0
+    c12 = 7.0 / 15.0
+    c13 = 11.0 / 70.0
+    c21 = 7.0 / 20.0
+    c22 = 9.0 / 28.0
+    c23 = 263.0 / 210.0
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:244]: Nope no clue would it not clear in the idl where they come from.

File: sunxspex/emission.py [hunk 13]
```diff
@@ -0,0 +1,814 @@
...
+    # Compute the Elwert factor.
+    a1 = alpha * z * e1 / p1
+    a2 = alpha * z * e2 / p2
+
+    fe = (a2 / a1) * (1.0 - np.exp(-2.0 * np.pi * a1)) / (1.0 - np.exp(-2.0 * np.pi * a2))
+
+    # Compute the differential cross section (units cm^2).
+    cross_section = twoar02 * fe * crtmp
+
+    return cross_section
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:283]: Same deal as above 😕

File: sunxspex/emission.py [hunk 14]
```diff
@@ -0,0 +1,814 @@
...
+        return electron_dist.density(electron_energy) * brem_cross * pc / collision_loss / gamma
+    elif model == 'thin-target':
+        if efd:
+            # if electron flux distribution is assumed (default)
+            return electron_dist.flux(electron_energy) * brem_cross * (mc2 / clight)
+        else:
+            # if electron density distribution is assumed
+            # n_e * sigma * mc2 * (v / c)
+            # TODO this is the same as IDL version but doesn't make sense as units are different?
+            return electron_dist.flux(electron_energy) * brem_cross * pc / gamma
```
[the remark this answers is not in the document]
Comment (samaloney) [reply sunxspex/emission.py:349]: The first case call the `density` method or at least it should but the second two are repeated

File: sunxspex/emission.py (hunk 3 above)
Comment (samaloney) [reply sunxspex/emission.py:579]: Yea I think you are right

File: sunxspex/emission.py (hunk 2 above)
Comment (samaloney) [reply sunxspex/emission.py:581]: It's 1-d so could probably just be `if P1:`

**Pull request merged**